src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts

```diff
@@ -0,0 +1,309 @@
+    } else {
+      this.previousHours = this.hours.slice(hoursIndex - 3, hoursIndex);
+    }
+    this.nextHours = this.hours.slice(hoursIndex + 1, hoursIndex + 4);
+  }
+
+  calculateSurroundingMinutes() {
+    const minutesIndex = Number(this.timeMinutes);
+    if (minutesIndex <= 3) {
+      this.previousMinutes = this.minutes.slice(0, minutesIndex);
+    } else {
+      this.previousMinutes = this.minutes.slice(minutesIndex - 3, minutesIndex);
+    }
+    this.nextMinutes = this.minutes.slice(minutesIndex + 1, minutesIndex + 4);
+  }
+
+  setDisabled() {
+    if (this.disabled) {
+      this.control?.disable();
+    } else {
+      this.control?.enable();
+    }
+  }
+
+  callbackFunction(e) {
+    this.userInput.emit(e);
+  }
+
+  blurCallBackFunction(e) {
+    if (!e.relatedTarget?.classList?.contains('clock-image')) {
+      this.removeFocusStyling();
+    }
+    this.blur.emit(e);
+  }
+
+  keyupCallBackFunction(e) {
+    this.keyup.emit(e);
+  }
+
+  stopPropagation(e) {
+    e.stopPropagation();
+  }
+
+  addFocusStyling() {
+    document.getElementById(`time-container-${this.id}`).classList.add('time-container-focus');
+  }
+
+  removeFocusStyling() {
+    document.getElementById(`time-container-${this.id}`).classList.remove('time-container-focus');
+  }
+
+  addValidationError = () => {
+    this.announceCurrentTime(true);
+    this.removeFocusStyling();
+    document.getElementById(`time-container-${this.id}`).classList.add('time-container-error-shake');
+    setTimeout(() => {
+      this.removeValidationError();
+    }, 5000);
+  };
+  removeValidationError = () => {
+    document.getElementById(`time-container-${this.id}`).classList.remove('time-container-error-shake');
+    this.addFocusStyling();
+  };
+
+  announceCurrentTime(hasError = false) {
+    this.screenReaderAnnouncement = `${hasError ? 'Invalid time provided' : ''}. The current time value is ${
+      this.control.value
+    }`;
+  }
+
+  /*
+  this function is called when the user hits enter or when they leave the input
+  and by the input after a debounce time
+
+   don't accept the input if there's
+   - a colon but it isn't in the 2nd or 3rd spot
+   - a colon but only one minute digit
+   - no colon but length isn't 3 or 4 digits
+   - an hour value greater than 23
+   - a minute value greater than 59
+  */
+  formatAndSetTime(event) {
+    const verifiedTime = /^(?:[01][0-9]|2[0-3]):[0-5][0-9]$/;
+    const input: string = event.target.value;
+    const colonIndex = input.indexOf(':');
+    let formattedInput;
+    let hours;
+    let minutes;
+    // check to see if the colon is in the 2nd or 3rd spot i.e. H:MM or HH:MM
+    if ([1, 2].includes(colonIndex)) {
+      [hours, minutes] = input.split(':');
+      hours = hours.padStart(2, '0');
+      if (minutes.length === 2) {
+        formattedInput = `${hours}:${minutes}`;
+      }
+    }
+    // if colon isn't in 2nd or 3rd spot, then we don't allow it anywhere else
```

Comment:
👍 comments explaining the "why" near the code in question, not just in the block at the top